Yuhsuan/1397 fitting enhancement

YuHsuan-Hwang opened the pull request:
**Description**

Closes #1397 remaining enhancement items in the image fitting feature. Requiring backend PR https://github.com/CARTAvis/carta-backend/pull/1237 and protobuf PR https://github.com/CARTAvis/carta-protobuf/pull/81.

Changes
 - Input panel enhancements
   - Support setting the fitting region. Options include FOV, entire image, and closed regions. The default option is FOV.
   - Support setting parameters in wcs coordinate ~(methods implemented, UI still in progress)~.
     - The "(px)" unit changes with the coordinate mode.
     - A info string is placed below the inputs to show the value of the other coordinate. If values are not set yet, we show "WCS/Image: -". If only x is set, we show "WCS: (VALUE, null)" or "Image: (VALUE pix, NaN pix)", and vise versa.
     - Note that: when setting center X, if center Y is not set yet, center Y will be initialized with the wcs coordinate of the origin (0, 0), and vise versa. Or if center Y is set and X is not, center Y will be updated with the wcs coordinate of (0, center Y), and vise versa.
   - Support setting a background offset as a constant or a free fitting parameter. The default is 0 and fixed.
   - Support different solver options. The default is Cholesky. See [the document](https://www.gnu.org/software/gsl/doc/html/nls.html#c.gsl_multifit_nlinear_solver) for the details of the solvers. Modified cheloesky is not supported because it's not available in all gsl versions.
 - Result panel enhancements
   - In the fitting result, set the unit of fwhm x and y to be the same, as [the comment](https://github.com/CARTAvis/carta-frontend/pull/1850#pullrequestreview-974794768) suggested.
   - Support creating ellipse regions from the fitting result, with default region name “Fitting result: Component #i” and dash length "2". A toast is shown after all the regions are successfully created.
 - UI design enhancements
   - Hovering on the unit of “Amplitube” and “Background” a while shows a tooltip.
    ![image](https://user-images.githubusercontent.com/43841102/216532813-16cb80ca-13ec-48c4-8caa-19041e6dd047.png)
   - The image input column is pinned when scrolling.
   - Added a resizer for adjusting the height of the input panel and the result panel.
   - Added a divider between component parameters and background parameter to show that background does not depend on different components.

Maintaniance
 - Fixed `ClearableNumericInputComponent` so that `inline={true}` works.

~ToDo~
 - [X] ~Add wcs input UI after the pan/zoom PR is merged.~

**Checklist**

For linked issues (if there are):
- [x] assignee and label added
- [x] ZenHub issue connection, board status, and estimate updated

For the pull request:
- [x] reviewers and assignee added
- [x] ZenHub estimate, milestone, and release (if needed) added
- [x] ~e2e test passing~ / corresponding fix added
- [x] changelog updated / ~no changelog update needed~
- [x] protobuf updated to the latest dev commit / ~no protobuf update needed~
- [x] `BackendService` unchanged / ~`BackendService` changed and corresponding ICD test fix added~

YuHsuan-Hwang committed:
enhance fwhm unit

YuHsuan-Hwang committed:
compare abs value

YuHsuan-Hwang committed:
add dropdown for region options

YuHsuan-Hwang committed:
implement creating regions from result

YuHsuan-Hwang committed:
fix bug

YuHsuan-Hwang committed:
add background offset input

YuHsuan-Hwang committed:
add solver options

YuHsuan-Hwang committed:
show toast after creating regions

YuHsuan-Hwang committed:
Merge branch 'dev' into yuhsuan/1397_fitting_enhancement

YuHsuan-Hwang committed:
fix eslint

YuHsuan-Hwang committed:
add checks when modifying params

YuHsuan-Hwang committed:
add wcs values

YuHsuan-Hwang committed:
add methods for wcs coord inputs

YuHsuan-Hwang committed:
pin image input

YuHsuan-Hwang committed:
implement resizer

YuHsuan-Hwang committed:
update center wcs with system

YuHsuan-Hwang committed:
show toast only if all regions are successfully created

YuHsuan-Hwang committed:
remove mcholesky option

kswang1029 commented:
@YuHsuan-Hwang if we define a region as the region for fitting and trigger the fit, we see the fitting is successful. However, if we then delete the region and select FOV as the region for fitting, there will be an error complaining ID not found.

In general, this draft looks excellent. Please convert both draft PRs to formal PRs. Code review may proceed.

YuHsuan-Hwang committed:
set region id to FOV when selected region is deleted

YuHsuan-Hwang committed:
add divider

YuHsuan-Hwang committed:
reformat

YuHsuan-Hwang committed:
fit with background

YuHsuan-Hwang committed:
update protobuf

YuHsuan-Hwang committed:
Merge branch 'dev' into yuhsuan/1397_fitting_enhancement

YuHsuan-Hwang committed:
add wcs input UI

YuHsuan-Hwang committed:
reformat

YuHsuan-Hwang commented:
New changes (also updated in the description):
* Fixed the region selection bug:
> if we define a region as the region for fitting and trigger the fit, we see the fitting is successful. However, if we then delete the region and select FOV as the region for fitting, there will be an error complaining ID not found.
* Added a divider between component parameters and background parameter to show that background does not depend on different components.
* Support background as a constant or a free fitting parameter. Added a button to fix/unfix background. Default is fixed.
* Added WCS input UI

protobuf PR is required for receiving the fitting result of the background as a free parameter.

YuHsuan-Hwang committed:
fix wcs inputs

YuHsuan-Hwang committed:
adjust px unit; add coord info string

YuHsuan-Hwang committed:
adjust set center wcs; adjust info string

YuHsuan-Hwang committed:
fix updating the input correctly when changing components with the input focused

YuHsuan-Hwang committed:
minor change

YuHsuan-Hwang commented:
New changes to the wcs inputs (also updated in the description):
- The "(px)" unit changes with the coordinate mode.
- A info string is placed below the inputs to show the value of the other coordinate. If values are not set yet, we show "WCS/Image: -". If only x is set, we show "WCS: (VALUE, null)" or "Image: (VALUE pix, NaN pix)", and vise versa.
- When setting center X, if center Y is set and X is not, center Y will be updated with the wcs coordinate of (0, center Y), and vise versa.
- Update the input correctly when changing components with the input focused

kswang1029 commented:
feature-wise this looks good. Please proceed code review.

kswang1029 reviewed (approved):
this looks good. 👍 

current e2e tests need a fix for xpath due to the new GUI. 
New e2e tests still need to be implement (a bit later on).
No regression found via current e2e tests.

YuHsuan-Hwang committed:
update region info in fitting log

YuHsuan-Hwang commented:
Fixed region info in the fitting log in the [comment](https://github.com/CARTAvis/carta-backend/pull/1237#issuecomment-1457518331).

kswang1029 reviewed (approved):
feature-wise this is looking very good. 👍 Please proceed code review.

YuHsuan-Hwang committed:
resolve conflict and merge dev

YuHsuan-Hwang committed:
fix text style

veggiesaurus commented on src/components/Dialogs/FittingDialog/FittingDialogComponent.tsx:
```diff
@@ -112,67 +149,110 @@ export class FittingDialogComponent extends React.Component {
+                                            stepSize={1}
+                                            max={fittingStore.components.length}
+                                            showTrackFill={false}
+                                            onChange={val => setTimeout(() => fittingStore.setSelectedComponentIndex(val - 1), 0)} // wait for onBlur events of the inputs
```
is there no better way to handle this? :thinking:

YuHsuan-Hwang replied:
@veggiesaurus Another way is to update the inputs with all keydown events instead of onBlur and enter key events. Currently I am reusing `CoordNumericInput`, so it's not easy. I will add a todo here.

YuHsuan-Hwang committed:
Merge branch 'dev' into yuhsuan/1397_fitting_enhancement

YuHsuan-Hwang committed:
add todo comment

YuHsuan-Hwang committed:
update protobuf

YuHsuan-Hwang committed:
update changelog

YuHsuan-Hwang committed:
update protobuf

YuHsuan-Hwang committed:
Merge branch 'dev' into yuhsuan/1397_fitting_enhancement

YuHsuan-Hwang committed:
update protobuf
